workbench/src/sidebar.c

```diff
@@ -0,0 +1,1154 @@
+ * belongs.
+ *
+ * @param context The location to store the selection context data in.
+ * @return TRUE if anything in the tree is selected, FALSE otherwise.
+ *
+ **/
+gboolean sidebar_file_view_get_selected_context(SIDEBAR_CONTEXT *context)
+{
+	gboolean has_parent;
+	guint dataid;
+	GtkTreeSelection *treesel;
+	GtkTreeModel *model;
+	GtkTreeIter current, parent;
+	gpointer data;
+
+	if (context == NULL)
+	{
+		return FALSE;
+	}
+	memset (context, 0, sizeof(*context));
+
+	treesel = gtk_tree_view_get_selection(GTK_TREE_VIEW(sidebar.file_view));
+	if (gtk_tree_selection_get_selected(treesel, &model, &current))
+	{
+		/* Search through the parents upwards until we find the project node.
+		   Save everthing that's interesting in callers variables... */
+		do
+		{
+			gtk_tree_model_get(model, &current, FILEVIEW_COLUMN_DATA_ID, &dataid, -1);
+			gtk_tree_model_get(model, &current, FILEVIEW_COLUMN_ASSIGNED_DATA_POINTER, &data, -1);
+			if (data != NULL)
+			{
+				switch (dataid)
+				{
+					case DATA_ID_WB_BOOKMARK:
+						context->wb_bookmark = data;
+					break;
+					case DATA_ID_PROJECT:
+						context->project = data;
+					break;
+					case DATA_ID_PRJ_BOOKMARK:
+						context->prj_bookmark = data;
+					break;
+					case DATA_ID_DIRECTORY:
+						context->directory = data;
+					break;
+					case DATA_ID_NO_DIRS:
+						/* Has not got any data. */
+					break;
+					case DATA_ID_FOLDER:
+						context->folder = data;
+					break;
+					case DATA_ID_FILE:
+						context->file = data;
+					break;
+				}
+			}
+
+			has_parent = gtk_tree_model_iter_parent(model,
+							&parent, &current);
+			current = parent;
+		}while (has_parent);
+
+		return TRUE;
+	}
+	return FALSE;
+}
+
+
+/** Setup the sidebar.
+ *
+ **/
+void sidebar_init(void)
+{
+	GtkWidget *scrollwin;
+	GtkCellRenderer *renderer;
+	GtkTreeViewColumn *column;
+	GtkTreeSelection *sel;
+	GList *focus_chain = NULL;
+
+	sidebar.file_view_vbox = gtk_vbox_new(FALSE, 0);
+
+	/**** label ****/
+	sidebar.file_view_label = gtk_label_new (_("No workbench opened."));
+	gtk_box_pack_start(GTK_BOX(sidebar.file_view_vbox), sidebar.file_view_label, FALSE, FALSE, 0);
+
+	/**** tree view ****/
+
+	sidebar.file_view = gtk_tree_view_new();
+	g_signal_connect(sidebar.file_view, "row-activated", (GCallback)sidebar_filew_view_on_row_activated, NULL);
+
+	sidebar.file_store = gtk_tree_store_new(FILEVIEW_N_COLUMNS, G_TYPE_ICON, G_TYPE_STRING, G_TYPE_UINT, G_TYPE_POINTER);
+	gtk_tree_view_set_model(GTK_TREE_VIEW(sidebar.file_view), GTK_TREE_MODEL(sidebar.file_store));
+
+	renderer = gtk_cell_renderer_pixbuf_new();
+	column = gtk_tree_view_column_new();
+	gtk_tree_view_column_pack_start(column, renderer, FALSE);
+	gtk_tree_view_column_add_attribute(column, renderer, "gicon", FILEVIEW_COLUMN_ICON);
+
+	renderer = gtk_cell_renderer_text_new();
+	gtk_tree_view_column_pack_start(column, renderer, TRUE);
+	gtk_tree_view_column_add_attribute(column, renderer, "markup", FILEVIEW_COLUMN_NAME);
```

Comment:
you should use `"text"`" instead of `"markup"` as you pass raw filenames (and don't seem to need the markup features), otherwise it'll break with filenames containing `&`s, `<`s and the like.